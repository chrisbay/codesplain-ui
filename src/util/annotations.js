import _ from 'lodash';

export const getAnnotatedLines = (annotations) => {
  return _.sortBy(_.keys(annotations).map(key => Number(key)));
};

export const getPreviousAnnotation = (annotations, displayedLineNumber) => {
  const annotatedLines = getAnnotatedLines(annotations);
  if (_.head(annotatedLines) === displayedLineNumber) {
    // First annotation is the one being displayed so there isn't another
    // annotation before this one; return undefined
    return undefined;
  }
  // Get the index of the currently displayed annotation in the annotatedLines
  // array and subtract one to get index of the previous annotation
  const index = _.sortedIndexOf(annotatedLines, displayedLineNumber) - 1;
  return annotations[annotatedLines[index]];
}

export const getNextAnnotation = (annotations, displayedLineNumber) => {
  const annotatedLines = getAnnotatedLines(annotations);
  if (_.last(annotatedLines) === displayedLineNumber) {
    // Last annotation is the one being displayed so there isn't another
    // annotation after this one; return undefined
    return undefined;
  }
  // Get the index of the currently displayed annotation in the annotatedLines
  // array and add one to get index of the next annotation
  const index = _.sortedIndexOf(annotatedLines, displayedLineNumber) + 1;
  return annotations[annotatedLines[index]];
}
