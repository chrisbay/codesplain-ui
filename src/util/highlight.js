import { styleRegion } from '../components/SnippetArea';

const ignoredTokens = [
  'suite',
  'file_input',
  'simple_stmt',
  'trailed_atom',
  'trailer',
]

const tokenColors = {
  'for_stmt' :    '#F7ABAB',
  'if_stmt':      '#C23B22',
  'number':       '#DEA5A4',
  'expr':         '#F0ABF7',
  'str':          '#CAABF7',
  'atom':         '#ABDBF7',
  'expr_stmt':    '#BEDA0B',
  'arglist':      '#F7CDAB',
  'argument':     '#EAF7AB',
  'integer':      '#ABF7C6',
  'funcdef':      '#AEC6CF',
  'parameters':   '#FFB347',
  'classdef':     '#03C03C',
  'try_stmt':     '#966FD6',
  'pass_stmt':    '#FDFD96',
  'except_clause':'#779ECB',
}

function getColor(type) {
  return tokenColors[type];
}

export function highlight(snippet, node, codeMirrorRef, filters, parentColor) {
  let color = parentColor;
  if (parentColor === undefined) color = 'transparent';

  // If we aren't ignoring this token...
  if (ignoredTokens.indexOf(node.type) === -1) {
    color = getColor(node.type); // Get the color for this token's type

    // If this token has no color 
    if (!color) {
      color = 'inherit';
      console.warn(`token "${node.type}" has no color specified!`);
    }

    // If this token's filter is not selected
    if (!filters[node.type].selected) {
      color = parentColor;
    }

    // Apply the background color CSS to this token
    styleRegion(
      codeMirrorRef, 
      node.begin, 
      node.end, 
      `background-color: ${color};`
    );
  }

  // Highlight all children of this token
  node.children.forEach(child => {
    if (child === Object(child)) 
      highlight(snippet, child, codeMirrorRef, filters, color);
  });
}
