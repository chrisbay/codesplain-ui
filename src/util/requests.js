const API_URL = process.env.REACT_APP_API_URL;

/* Construct the endpoint to make a REST request to. Only the username is
 * field is required; the snippetId will be left blank for POST requests,
 * and can be supplied when needed (say, for PUT requests).
 *
 */
export const makeSaveEndpointUrl = (username, snippetId = '') => {
  if (snippetId) {
    return `${API_URL}/users/${username}/snippets/${snippetId}`;
  }
  return `${API_URL}/users/${username}/snippets`;
}
