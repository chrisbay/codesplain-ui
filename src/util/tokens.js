const prettyTokenNameMap = {
  'for_stmt' :    'For Loops',
  'if_stmt':      'If Statements',
  'number':       'Numbers',
  'expr':         'Expressions',
  'str':          'Strings',
  'atom':         'Atoms',
  'expr_stmt':    'Expression Statements',
  'arglist':      'Argument Lists',
  'argument':     'Arguments',
  'integer':      'Integers',
  'funcdef':      'Function Definitions',
  'parameters':   'Parameters',
  'classdef':     'Class Definitions',
  'try_stmt':     'Try Statements',
  'pass_stmt':    'Pass Statements',
  'except_clause':'Except Clauses',
}

/* Maps AST token types to pretty strings for displaying */
export function getPrettyTokenName(token) {
  return prettyTokenNameMap[token];
}

/* (Recursively) populates map with each token type's occurence count. */
export function getTokenCount(node, map) {
  if (node.type === undefined) return;
  if (map[node.type] === undefined)
    map[node.type] = 1;
  else 
    map[node.type] += 1;
  node.children.forEach(child => getTokenCount(child, map));
}
