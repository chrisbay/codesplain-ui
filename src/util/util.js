/*
Returns a function that will convert a character index to an object
of its row and column in the context of the passed-in snippet.

The returned object is of the following form: { line: <int>, ch: <int> }

Regards the first row in a snippet as row 0, and the first column of each
row as column 0.
*/
export function getIndexToRowColConverter(snippet) {
  // Make array containing the length of each lines of code in snippet
  const lines = snippet.split('\n').map(l => l.length + 1); // +1 to accomodate '\n' that was lost when we split the snippet
  return (index) => {
    let row, col, lenPrevRows;
    row = lenPrevRows = 0;
    while (index > lines[row] + lenPrevRows)
      lenPrevRows += lines[row++];
    col = index - lenPrevRows;
    return { line: row, ch: col }
  }
}
