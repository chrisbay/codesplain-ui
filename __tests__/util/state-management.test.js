import * as utils from '../../src/util/state-management';

describe('util: state-management', () => {
  describe('defaultStateValues', () => {
    it('matches snapshot', () => {
      expect(utils.defaultStateValues).toMatchSnapshot();
    });
  });
  describe('setDefaults', () => {
    it('sets the default state values', () => {
      expect(utils.setDefaults({})).toEqual(utils.defaultStateValues);
    });
    it('does not override pre-existing state properties', () => {
      const snippetLanguage = 'java';
      const state = {
        snippetLanguage,
      };
      expect(utils.setDefaults(state).snippetLanguage).toEqual(snippetLanguage);
    });
  });
});
