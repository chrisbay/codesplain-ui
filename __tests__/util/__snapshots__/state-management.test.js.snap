exports[`util: state-management defaultStateValues matches snapshot 1`] = `
Object {
  "snippetLanguage": "python3",
}
`;
